PR-Review #1

Hyune-c opened the pull request:
모든 소스를 `master` 로 최신화하고 `be-review` 를 통해 리뷰를 요청했습니다

### Work Log
- https://github.com/Hyune-c/TIL/tree/master/Toy%20Project/Raider50G

### 기능 설명
- Issue 에 Domain 을 기준으로 설명해 놓았습니다

### 리뷰시 요청 내용
- 리뷰를 받게될 기회가 이렇게 빨리 생길지 몰라서 소스 내 주석을 달지 못했습니다.. ㅠ
- 시그리드는 평소에 이런 실 개발 외 활동도 많이 하시는 것 같아 가능하시다면 기능 외 내용도 리뷰를 부탁드립니다
  - 코드가 잘 읽히는가 & 말이 되는가?
  - Issue 에 내용이 잘 정리되어있는가?
  - Repository 가 읽기 쉬운가? (특히 README)
  - Work Log 가 유용한가?

Hyune-c committed:
chore: spring-boot-starter-data-jpa 추가

- spring-boot-starter-data-jpa 추가

Hyune-c committed:
feat: Booking 만들기

- Entity 생성
- Enum type 추가
  - ClassType, RaidType

#6

Hyune-c committed:
refactor: Booking Entity 개선

- BaseInfo, Raider 의 추가로 Booking Entity 를 간결하게 만듬

#6

Hyune-c committed:
chore: gson 추가

- json 객체 직렬화를 위해 추가

Hyune-c committed:
refactor: Booking 의 Embedded 객체 개선

- 생성자 추가
  - 기본 생성자
  - 입력 값에 대한 초기 생성자와 초기 값 설정
  - Builder 설정
- Swagger 상에 드러나지 않아도 되는 파라미터에 대한 설정

#6

Hyune-c committed:
feat: BookingCommand 객체 생성

- 예약을 위한 입력 객체

#6

Hyune-c committed:
refactor: RaidInfo 에 테스트 관련 메소드 추가

- getTestRaidInfo()
  - 외부에서의 입력 값과 같은 형태로 설정하기 위한 메소드

Hyune-c committed:
test: BookingCommand 의 생성 테스트

- BookingCommand 의 생성 테스트
  - DDD 에 기반한 테스트

Hyune-c committed:
refactor: 사용하지 않는 클래스 모두 삭제

- Phase 2 로 넘어가면서 프로젝트 구조를 다시 설계하게 되어 클래스 정리를 하였습니다

Hyune-c committed:
refactor: 사용하지 않는 클래스 모두 삭제

- Phase 2 로 넘어가면서 프로젝트 구조를 다시 설계하게 되어 클래스 정리를 하였습니다

Hyune-c committed:
fix: 잘못된 import 수정

- 잘못된 import 수정

Hyune-c committed:
refactor: JsonUtil 정리

- 사용되지 않는 메소드 정리
- LocalDate 타입을 받을 수 있도록 설정

Hyune-c committed:
test: BookingCommand 의 생성 테스트

- given 절을 좀 더 세분화

Hyune-c committed:
chore: 설정 조정

- gson dependencies 제거
- git ignore 설정

Hyune-c committed:
refactor: exception 처리 단순화

- 표준 객체에서 String 으로 리턴하도록 조정

Hyune-c committed:
refactor: Entity 들을 어노테이션으로 정리

- Entity 들을 어노테이션으로 정리

Hyune-c committed:
refactor: DTO 를 어노테이션으로 정리

- DTO 를 어노테이션으로 정리

Hyune-c committed:
refactor: DevController 이름 변경과 테스트 케이스 개선

- DevBookingController 에서 BookingController 로 변경
- 테스트 케이스 개선

Hyune-c committed:
refactor: exception 처리 단순화

- 에러 표준 객체로 에러 클래스명과 에러 사유 응답

Hyune-c committed:
refactor: 프로젝트 구조 변경과 오타 수정

- 프로젝트 구조 변경

Hyune-c committed:
feat: Booking 을 DB 에 저장하는 기능 추가

- Booking 을 DB 에 저장하는 기능 추가

Hyune-c committed:
chore: querydsl 설정 추가

- querydsl 설정 추가

Hyune-c committed:
chore: querydsl 설정 추가

- querydsl 설정 추가

Hyune-c committed:
feat: find Booking 구현

- 해당 날짜에 취소되지 않은 예약만 가져오도록 구현

Hyune-c committed:
refactor: 프로젝트 구조 변경

- type 을 common 으로 이동
- Channel 을 enum 으로 변경

Hyune-c committed:
refactor: 프로젝트 구조 변경

- import 변경

Hyune-c committed:
feat: 예약 리스트 만들기 구현

- 예약 리스트 만들기 구현

Hyune-c committed:
feat: 예약 리스트를 특정 채널에 보내기 구현

- 만들어진 예약 리스트를 지정된 채널에 게시합니다

Hyune-c committed:
chore: build.gradle 정리

- build.gradle 정리

Hyune-c committed:
refactor: Property 파일 정리

- Property 파일 정리

Hyune-c committed:
Merge pull request #10 from Hyune-c/phase2/booking

Phase2/booking

Hyune-c committed:
chore: 설정 파일은 올리지 않게 변경

- 설정 파일은 올리지 않게 변경

Hyune-c committed:
chore: H2 Local 설정 추가

- H2 Local 설정 추가

Hyune-c committed:
feat: 초대 매크로 만들기

- /api/booking/invite-macro
  - raidDate 를 파라미터로 받아 매크로 스트링을 리턴합니다

Hyune-c committed:
refactor: 초대 매크로 만들기를 객체 내부 메소드로 수정

- 초대 매크로 만들기를 객체 내부 메소드로 수정

Hyune-c committed:
refactor: Repository 의 find() 명 변경 & 문구 변경

- find() 에서 findAll() 로 변경
  - find() 는 단수를 리턴, findAll() 은 복수를 리턴하도록 정의
- makeBookingList() 의 title 을 좀 더 구체적으로 변경
- 테스트 케이스 문구 개선

Hyune-c committed:
feat: 예약 취소 기능 구현

- BookingCommand 를 통해 Booking 을 찾고 취소합니다
- 이미 취소된 예약인 경우 에러 발생

Hyune-c committed:
test: 테스트 케이스 정리

- cancel() 테스트 케이스 추가
- 테이스 케이스 문구를 한글로 정리하고 @Nested 사용으로 정리

Hyune-c committed:
refactor: RaidInfo 의 createdAt 을 LocalDateTime 으로 변경

- RaidInfo 의 createdAt 을 LocalDateTime 으로 변경

Hyune-c committed:
docs: README.md 수정

- 프로젝트 설명 추가

Hyune-c committed:
Merge branch 'master' into feature/booking

Hyune-c committed:
Merge pull request #11 from Hyune-c/feature/booking

Feature/booking

Hyune-c committed:
refactor: 중요하지 않은 로직 개선

- ClassType
  - name 을 한개만 가지도록 개선
- Booking
  - 내부 객체로 접근하지 않고 값을 가져올 수 있도록 getRaiderId(), getClassType() 추가
- ChannelService
  - makeBookingList() 에서 createBookingList() 로 이름 변경
  - 메소드 분리

Hyune-c committed:
test: ChannelServiceTest 추가

- makeBookingList(): 예약 인원 리스트 만들기

Hyune-c committed:
chore: Swagger 제거, WebFlux 설정

- WebFlux 설정
  - RestTemplate 가 Deprecated 예정이기에 기술 전환
- Swagger 삭제
  - 3.0.0 미만에서 WebFlux 와 호환되지 않아 삭제

Hyune-c committed:
refactor: 커스텀 Getter 에 @JsonIgnore 추가

- 커스텀 Getter 의 추가로 Json Response 가 의도대로 내려가지 않아 추가

Hyune-c committed:
refactor: Booking cancel() 버그 픽스

- Booking cancel()
  - EntityManager 에 등록되지 않아 영속성 관리가 되지 않았습니다

Hyune-c committed:
refactor: sendBookingList() 기능 개선

- WebClient 적용
  - return 형태를 Mono<Object> 로 개선
- createBookingList() 에서 createBookingListString() 으로 메소드명 변경

Hyune-c committed:
test: `예약 리스트 보내기` 테스트 케이스 추가

- @SpringBootTest 를 RANDOM_PORT 로 작동하도록 적용
- sendBookingList()
  - Discord 에서 온 응답을 바로 전달하도록 수정

Hyune-c committed:
refactor: `예약 리스트 보내기` 개선

- createBookingListString()
  - public 으로 변경되며 bookingListString 를 만듭니다
- sendBookingList()
  - 메소드 명에 맞도록 매개변수에 bookingListString 를 받습니다
- 메소드의 개선으로 테스트에서 bookings 를 주입하지 않아도 됩니다

Hyune-c committed:
refactor: `예약 리스트 보내기` 개선

- @PathVariable 타입 변경
  - 기존 String 으로 받아 Enum 으로 변환하던 것을 처음부터 Enum 으로 받을 수 있도록 변경

Hyune-c committed:
refactor: Booking Entity 변경

- cancel, createdAt 위치 변경
  - 기존 RaidInfo 객체에서 Booking 객체가 직접 관리하도록 위치 변경

Hyune-c committed:
refactor: RaidInfo 객체의 of() 삭제

- RaidInfo 객체의 필드 위치 변경으로 인해 of() 삭제

Hyune-c committed:
refactor: DungeonType 으로 Channel 병합

- 공격대에 참여하기 위해 신청하는 것은 채널이 기준이 아닌 던전이 기준이기에 하나로 합쳤습니다

- 프로젝트 설명 추가

Hyune-c committed:
feat: BookingList 객체 추가

- Booking 들을 묶어줄 객체가 필요하여 추가하였습니다

Hyune-c committed:
refactor: `예약 리스트 게시하기` 기능 개선

- postMessages() 메소드 추가
  - 지정된 채널에 jsonBody 를 작성합니다
- BookingList 객체의 생성으로 해당 객체를 만들고, 게시할 json 을 createBookingSheet() 로 형태로 받도록 기능을 개선했습니다

Hyune-c committed:
refactor: `예약 리스트 게시하기` 기능 개선

- postBookingList() 와 postMessages() 의 관계를 좀 더 말이 되게 개선
  - postMessages() 는 게시할 채널과 body 만 알면 되기에 변수 타입과 이름을 개선했습니다

Hyune-c committed:
test: BookingList 객체 생성에 따라 테스트 구조 개선

- BookingList 객체 생성에 따라 테스트 구조 개선

Hyune-c committed:
Merge pull request #12 from Hyune-c/feat/channel

Feat/channel

realsigridjin commented:
@Hyune-c 안녕하세요. 리뷰어로 저를 지정해주시고 디엠 부탁드립니다.

Hyune-c commented:
> @Hyune-c 안녕하세요. 리뷰어로 저를 지정해주시고 디엠 부탁드립니다.

sigridjineth reviewed (approved):
**[리뷰어를 스스로 지정할 수 있다]** 테스트

Hyune-c committed:
refactor: BookingList 개선

- DTO 로 위치 변경
- bookingsSheet 필드를 만들어 BookingList 의 역할만 할 수 있도록 합니다

Hyune-c committed:
refactor: BookingList 개선

- Stream 을 통해 함수화하였습니다
- getContents()
  - 예약된 인원이 존재하는 직업과 존재핮 않는 직업을 나눠서 변수화하였습니다

Hyune-c committed:
refactor: ClassType 에 최소, 최대 인워 수 추가

- ClassType 에 최소, 최대 인워 수 추가

Hyune-c committed:
refactor: BookingList 개선

- MultiValueMap 의 key 값을 ClassType 으로 개선
- 메소드 병합

Hyune-c committed:
test: test 개선

- TestJsonData 추가
- BookingList POJO test 추가

Hyune-c committed:
refactor: BookingList 오타 수정

- BookingList 오타 수정

Hyune-c committed:
Merge pull request #14 from Hyune-c/feat/channel

Feat/channel

Hyune-c committed:
feat: 채널에서 메세지 가져오기 기능 추가

- DiscordMessage, DiscordUser
  - Discord 에서 제공하는 Entity 를 implement 하였습니다
- getMessages()
  - 채널에서 메세지 가져오기

Hyune-c committed:
refactor: 채널에서 메세지 가져오기 기능 개선

- DiscordMessage
  - 필드 타입을 DiscordUser 에서 User 로 변경
  - getUserAuthor() 구현
- DiscordUser
  - getName() 구현
- getMessages()
  - 조건문 개선

Hyune-c committed:
refactor: DiscordMessage 에 createdAt 필드 추가

- DiscordMessage 에 createdAt 필드 추가

Hyune-c committed:
refactor: DiscordMessage 에 id 필드 추가

- DiscordMessage 에 id 필드 추가

Hyune-c committed:
refactor: WebClientService 로 WebClient 관련 로직 분리

- WebClientService
  - API 를 호출하는 모든 메소드를 분리
  - 필드 타입을 DiscordUser 에서 User 로 변경
  - getUserAuthor() 구현
- DiscordUser
  - getName() 구현
- getMessages()
  - 채널에서 메세지 가져오기

Hyune-c committed:
refactor: 코드 개선 & 실수 수정

- DiscordMessage
  - id 의 타입을 String 에서 long 으로 수정
- Test
  - ChannelServiceTest 를 WebClientServiceTest 로 수정

Hyune-c committed:
test: 테스트 케이스 개선

- 테스트 케이스 개선

Hyune-c committed:
refactor: WebClientService 개선

- 반복되는 코드 개선
  - uriFunction, headersConsumer 를 통해 반복 코드 제거

Hyune-c committed:
feat: 수동 `예약 리스트 지우기` 구현

- 기본 코드는 작성했지만, 1개의 쓰레드에서 Block 을 하여 정상 작동하지 않습니다

Hyune-c committed:
refactor: DiscordProperty 설정 추가와 오타 수정

- DiscordProperty
  - userAgent 추가

Hyune-c committed:
refactor: defaultHeaders 로 header 통합

- defaultHeaders 로 header 통합

Hyune-c committed:
refactor: makeInviteMacro() 를 만들면서 특정 유저의 이름은 제외하는 로직 추가

- 공대장의 이름은 제외하는 로직 추가

Hyune-c committed:
refactor: Raider Entity 의 raiderId 필드를 userName 으로 변경

- Raider Entity 의 raiderId 필드를 userName 으로 변경

Hyune-c committed:
refactor: exceptUserName 이 필수 값이 아니어도 되도록 명시

- exceptUserName 이 필수 값이 아니어도 되도록 명시

Hyune-c committed:
docs: README.md 개선

- WorkLog 추가

Hyune-c committed:
Merge pull request #15 from Hyune-c/feat/channel

Feat/channel

Hyune-c committed:
Update README.md

realsigridjin commented on build.gradle:
```diff
@@ -13,26 +14,56 @@ repositories {
 dependencies {
-    implementation 'org.springframework.boot:spring-boot-starter-web'
-    implementation 'org.springframework.boot:spring-boot-starter'
+    implementation group: 'com.querydsl', name: 'querydsl-jpa', version: '4.3.1'
```
오. 궁금했던 query dsl.
여기서는 RAW JPA를 사용하지는 않으신건가요?

Hyune-c replied:
아뇨. 사용했습니다.
`spring-boot-starter-data-jpa` dependency 로 implementation 했습니다.

realsigridjin commented on build.gradle:
```diff
@@ -13,26 +14,56 @@ repositories {
     useJUnitPlatform()
 }
+
+def querydslSrcDir = 'src/main/generated'
```
def는 무슨 의미를 갖고 있나요?

Hyune-c replied:
Querydsl 을 사용하려면 Entity 들을 QClass 라는 형태로 만들어 주어야 되더라구요.
그에 관련된 설정으로 위 경로에 QClass 들이 모여있습니다.

realsigridjin replied:
@Hyune-c `'src/main'` 에 `generated` 라는 폴더가 없어 보이는데요?

Hyune-c replied:
어 그러게요.. 뭐지..
근데 아래 코멘트 달으신 것 중에 보면 QClass 들이 잡히긴 해요.

![image](https://user-images.githubusercontent.com/55722186/88639639-f4284200-d0f7-11ea-80ad-c087816dfa1c.png)

realsigridjin commented on build.gradle:
```diff
@@ -13,26 +14,56 @@ repositories {
     implementation group: 'com.googlecode.json-simple', name: 'json-simple', version: '1.1.1'
     implementation group: 'com.google.code.gson', name: 'gson', version: '2.8.6'
+
+    implementation 'joda-time:joda-time:2.9.4'
```
joda time 사용 👍

realsigridjin commented on src/main/generated/com/hyune/raider50g/domain/booking/QRaider.java:
```diff
@@ -0,0 +1,39 @@
+
+    public static final QRaider raider = new QRaider("raider");
+
+    public final EnumPath<com.hyune.raider50g.common.type.ClassType> classType = createEnum("classType", com.hyune.raider50g.common.type.ClassType.class);
```
굳이 import static으로 할 필요는 없을 것 같아요.

Hyune-c replied:
QClass 는 Querydsl 사용시 자동 생성되는 것이라 내부 내용은 잘 못 보았네요..
이번 기회에 좀 봐야겠습니다.

realsigridjin commented on src/main/generated/com/hyune/raider50g/domain/booking/QRaider.java:
```diff
@@ -0,0 +1,39 @@
+@Generated("com.querydsl.codegen.EmbeddableSerializer")
+public class QRaider extends BeanPath<Raider> {
+
+    private static final long serialVersionUID = -48997379L;
```
고정 값인가요?

Hyune-c replied:
QClass 는 Querydsl 사용시 자동 생성되는 것이라 내부 내용은 잘 못 보았네요..
이번 기회에 좀 봐야겠습니다.

realsigridjin commented on src/main/java/com/hyune/raider50g/common/type/ClassType.java:
```diff
@@ -0,0 +1,19 @@
+@AllArgsConstructor
+public enum ClassType {
+
+  DRUID("DRUID", "드루", 2, 3), HUNTER("HUNTER", "냥꾼", 4, 5), MAGE("MAGE", "법사", 6, 7), PRIEST(
```
가독성 수정 추천드립니다.
```
DRUID("DRUID", "드루", 2, 3), 
HUNTER("HUNTER", "냥꾼", 4, 5)
...
```

realsigridjin commented on src/main/java/com/hyune/raider50g/config/H2ServerConfiguration.java:
```diff
@@ -0,0 +1,18 @@
+
+@Profile("local")
+@Configuration
+public class H2ServerConfiguration {
```
H2 프로파일 구분 👍

realsigridjin commented on src/main/java/com/hyune/raider50g/config/property/DiscordProperty.java:
```diff
@@ -0,0 +1,17 @@
+@Getter
+@Setter
+@Configuration
+@ConfigurationProperties(prefix = "discord")
```
discord 프로퍼티 구분 👍

realsigridjin commented on src/main/java/com/hyune/raider50g/domain/booking/Booking.java:
```diff
@@ -0,0 +1,69 @@
+  @GeneratedValue(strategy = GenerationType.IDENTITY)
+  private Long id;
+
+  @Builder.Default
```
빌더 어노테이션은 생성자 쪽에서 넣어주세요. 그래야 의도하지 않은 사이드 이펙트를 막을 수 있습니다.
https://cheese10yun.github.io/lombok/

realsigridjin commented on src/main/java/com/hyune/raider50g/domain/booking/Booking.java:
```diff
@@ -0,0 +1,69 @@
+  @Embedded
+  private RaidInfo raidInfo;
+
+  @Embedded
```
적절한 ```@Embedded``` 사용이 좋네요. 👍

realsigridjin commented on src/main/java/com/hyune/raider50g/domain/booking/Booking.java:
```diff
@@ -0,0 +1,69 @@
+
+  @Builder.Default
+  @Column(nullable = false)
+  private Boolean cancel = false;
```
Wrapper Class를 사용한 이유는요?

Hyune-c replied:
Booking 은 Entity 이기에 DB 와 직접 연결 됩니다.
그리고 DB 에 연결되는 값은 Wrapper Class 를 사용하는 것이 더 안전하다고 알고 있습니다.

realsigridjin replied:
네, 저도 그렇게 알고 있고 JPA에서 객체 매핑을 할 때에는 wrapper class를 사용하려고 합니다.
https://stackoverflow.com/questions/7506802/using-wrapper-integer-class-or-int-primitive-in-hibernate-mapping

> Use wrappers, make your life simple. Your data model should dictate this. You should be enforcing nullability in the database anyway. If they are nullable in the database, then use wrappers. If they are not nullable, and you use wrappers, then you'll get an exception if you try and insert a null into the database. If your data model doesn't dictate it, then go for a convention, use wrappers all of the time. That way people don't have to think, or decide that a value of 0 means null. I would also query your assertion that it would less performant. Have you measured this? I mean really measured it? When you're talking to a database, there are a lot more considerations than the difference between 16 bits and 32 bits. Just use the simple, consistent solution. Use wrappers everywhere, unless somebody gives you a very good reason (with accurate measured statistics) to do otherwise.

realsigridjin commented on src/main/java/com/hyune/raider50g/domain/booking/Raider.java:
```diff
@@ -0,0 +1,31 @@
+  private String userName;
+
+  public String inviteMacro() {
+    return "/초대 " + userName;
```
```/초대``` 이게 뭔가요?

Hyune-c replied:
게임 내 초대 매크로를 만들기 위한 형태입니다.
ex) /초대 전사1

realsigridjin commented on src/main/java/com/hyune/raider50g/domain/booking/dto/BookingList.java:
```diff
@@ -0,0 +1,66 @@
+
+  private Integer bookingCount() {
+    return classListMap.keySet().stream()
+        .map(key -> classListMap.get(key).size())
```
classListMap을 일급 콜렉션으로 묶어보세요.
객체지향의 원칙에서 디미터 법칙을 지켜주시는 것이 클린코드로 가는 길이라고 생각합니다.
객체는 외부 객체가 무엇을 하는지 모르도록 해야 합니다.
https://github.com/woowacourse/javable/pull/89/files

Hyune-c replied:
일급 컬렉션은 이름만 알고 있었는데, 이번에 적용해보면 좋을 것 같습니다.
개선해보겠습니다!

realsigridjin commented on src/main/java/com/hyune/raider50g/domain/booking/dto/BookingList.java:
```diff
@@ -0,0 +1,66 @@
+  }
+
+  private String makeClassLine(ClassType key) {
+    return (Objects.isNull(classListMap.get(key)))
```
계속 get으로 꺼내고 있네요. 객체에서 정보를 깨내지 말고 일을 시키도록 메시지를 보내보세요.

Hyune-c replied:
classListMap 를 일급 컬렉션으로 만드는 형태를 통해 개선할 수 있을 듯 합니다.
다음 버전에서는 그렇게 되도록 만들겠습니다.

realsigridjin commented on src/main/java/com/hyune/raider50g/domain/booking/dto/BookingList.java:
```diff
@@ -0,0 +1,66 @@
+        // 예약이 없는 직업군이면
+        ? String.format("%s (0/%d)", key.getName(), key.getMaxCount())
+        // 예약이 있는 직업군이면
+        : classListMap.get(key).stream().collect(Collectors.joining(
```
메소드 하나가 너무 여러 일을 하고 있는데요.
메소드를 분리해주세요. 현란한 Collectors 사용은 인상적이네요 👍

Hyune-c replied:
네. 개선 방법을 생각해보겠습니다.

realsigridjin commented on src/main/java/com/hyune/raider50g/domain/booking/dto/BookingList.java:
```diff
@@ -0,0 +1,66 @@
+        , raidInfo.getDungeonType().getName()
+        , raidInfo.getRaidDate()
+        , bookingCount());
+    String category = "직업 현재/최대\t예약 인원\t(2~3탱전 11힐)";
```
카테고리 메시지는 static하게 관리할 수 있지 않을까요?

Hyune-c replied:
네. 분리해 보겠습니다.

realsigridjin commented on src/main/java/com/hyune/raider50g/domain/booking/dto/BookingList.java:
```diff
@@ -0,0 +1,66 @@
+
+  public String createBookingSheet() {
+    String title = String.format("%s\t%s (일) PM 19:00\t%d/40"
+        , raidInfo.getDungeonType().getName()
```
여기도 title 하나를 만들기 위해 너무 많이 꺼내고 있네요. 리팩토링의 대상입니다.

Hyune-c replied:
네. 분리해 보겠습니다.

realsigridjin commented on src/main/java/com/hyune/raider50g/domain/channel/DiscordMessage.java:
```diff
@@ -0,0 +1,278 @@
+
+  @Override
+  public MessageAuthor getAuthor() {
+    return null;
```
음... 계속 null 값을 반환해야 하는 경우가 너무 많이 생기네요.
인터페이스를 사용하는 것이 맞는지 재고해주세요. 경우에 따라 추상 클래스를 사용하는 것이 더 효율적일 수도 있습니다.

Hyune-c replied:
`DiscordMessage` 는 `discord 에서 제공하는 Message` 를 implement 하면서, 필요한 필드만 가져와서 위와 같은 문제가 생겼습니다.
아마 개발이 진행되면서 해당 필드가 실제로 쓰인다면 점점 채워져갈 것으로 생각됩니다.

jypthemiracle commented on src/main/java/com/hyune/raider50g/domain/channel/DiscordMessage.java:
```diff
@@ -0,0 +1,278 @@
+
+  @Override
+  public CompletableFuture<Void> removeOwnReactionByEmoji(String unicodeEmoji) {
+    return null;
```
하나님...

Hyune-c replied:
`DiscordMessage` 는 `discord 에서 제공하는 Message` 를 implement 하면서, 필요한 필드만 가져와서 위와 같은 문제가 생겼습니다.
아마 개발이 진행되면서 해당 필드가 실제로 쓰인다면 점점 채워져갈 것으로 생각됩니다.

realsigridjin commented on src/main/java/com/hyune/raider50g/exception/CustomAdvice.java:
```diff
@@ -13,19 +10,7 @@
-  public ApiResponse handleFailedBookingException(FailedBookingException e) {
-    return ApiResponse.error(e.getMessage());
+  public ExceptionResponse handleException(Exception e) {
+    return ExceptionResponse.from(e);
```
왜 세부적으로 나눈 Custom Exception을 다 지우셨나요?

Hyune-c replied:
예외 처리에 신경쓰는 걸 후순위로 돌려서 원시적인 형태로 돌렸습니다.

realsigridjin commented on src/main/java/com/hyune/raider50g/exception/ExceptionResponse.java:
```diff
@@ -0,0 +1,22 @@
+public class ExceptionResponse {
+
+  String exceptionClass;
+  String message;
```
접근 생성자가 명시되지 않은 이유는요?

Hyune-c replied:
누락입니다..
수정하겠습니다!

realsigridjin commented on src/main/java/com/hyune/raider50g/exception/ExceptionResponse.java:
```diff
@@ -0,0 +1,22 @@
+
+  public static ExceptionResponse from(Exception e) {
+    return ExceptionResponse.builder()
+        .exceptionClass(e.getClass().toString())
```
그냥 Exception에서 ExceptionResponse를 build할 수 있지 않을까요?
굳이 DTO에서 별도의 build를 만들 필요는 없는 것 같아요.

realsigridjin replied:
**도전 질문**

만약에 도메인을 DTO로 만들어주는 부분이 있다고 해봅시다. 서비스 레이어에서 수행할 수는 없을까요? 분리해줄 수는 없을까요?
DTO 패턴을 활용하여 Assembler를 분리해주어도 좋을 것 같습니다.

martinfowler.com/eaaCatalog/dataTransferObject.html

realsigridjin commented on src/main/java/com/hyune/raider50g/repository/BookingRepository.java:
```diff
@@ -0,0 +1,51 @@
+public class BookingRepository {
+
+  @PersistenceContext
+  private EntityManager em;
```
RAW JPA를 사용하고 있군요.

realsigridjin commented on src/main/java/com/hyune/raider50g/repository/BookingRepository.java:
```diff
@@ -0,0 +1,51 @@
+  }
+
+  public Booking findOne(LocalDate findDate, String findUserName) {
+    JPAQueryFactory queryFactory = new JPAQueryFactory(em);
```
querydsl의 등장!

realsigridjin commented on src/main/java/com/hyune/raider50g/repository/BookingRepository.java:
```diff
@@ -0,0 +1,51 @@
+    JPAQueryFactory queryFactory = new JPAQueryFactory(em);
+    QBooking booking = QBooking.booking;
+
+    return queryFactory.selectFrom(booking)
```
직접 사전에 지정된 매개변수를 넣어주면 알아서 처리를 해주나보네요?

Hyune-c replied:
네 맞습니다.

realsigridjin commented on src/main/java/com/hyune/raider50g/service/BookingService.java:
```diff
@@ -0,0 +1,44 @@
+  public String makeInviteMacro(LocalDate findDate, String exceptUserName) {
+    return bookingRepository.findAll(findDate).stream()
+        .filter(booking -> !booking.getUserName().equals(exceptUserName))
+        .map(booking -> booking.getRaider().inviteMacro())
```
여기도 리팩토링의 냄새가 느껴지세요?

Hyune-c replied:
음.. 혹시 어떤 개선이 있으면 좋을까요?

realsigridjin replied:
객체에서 계속 무언가를 꺼내고 있네요. 꺼내지 말고 객체에 요청의 메시지를 보내도록 개선해보세요.

Hyune-c replied:
디미터에 관련된 얘기이신 것 같네요. 
좀 더 고민해보겠습니다.

realsigridjin commented on src/main/java/com/hyune/raider50g/service/WebClientService.java:
```diff
@@ -0,0 +1,114 @@
+        .pathSegment("channels", "{channelId}", "messages")
+        .build(channelId));
+
+    return webClient
```
webClient의 적절한 사용 👍

realsigridjin commented on src/main/java/com/hyune/raider50g/service/WebClientService.java:
```diff
@@ -0,0 +1,114 @@
+        .uri(uriFunction)
+        .accept(MediaType.APPLICATION_JSON)
+        .retrieve()
+        .bodyToMono(Object.class);
```
왜 Object에요?

Hyune-c replied:
Object 가 아닌 의미있는 객체를 넣어주는게 정상인데.. 실패했습니다 ㅠ

realsigridjin commented on src/main/java/com/hyune/raider50g/service/WebClientService.java:
```diff
@@ -0,0 +1,114 @@
+  }
+
+  public Mono<Object> deleteMessages(String channelId, long messageId) {
+    Function<UriBuilder, URI> uriFunction = (uriBuilder -> uriBuilder
```
적절한 Function 사용과 uriFunction 구현 좋네요 👍

realsigridjin commented on src/main/java/com/hyune/raider50g/service/WebClientService.java:
```diff
@@ -0,0 +1,114 @@
+    }
+
+    // Discord API 에서 온 응답을 파싱합니다
+    discordMessages = response.stream()
```
별도의 클래스로 빼주시죠.

realsigridjin commented on src/main/java/com/hyune/raider50g/service/WebClientService.java:
```diff
@@ -0,0 +1,114 @@
+@Service
+public class WebClientService {
+
+  private final WebClient webClient;
```
자. Mono를 쓰셨는데, 질문입니다.
혹시 Spring Reactive를 사용했음에도 여전히 Blocking API로 동작하고 있는 것은 아닌가요.
아래 글을 참고해보세요. 쓰레드가 하나일 수도 있습니다.
https://itstory.tk/entry/Spring-Webflux-JDBC%ED%98%B9%EC%9D%80-blocking-call-%ED%95%B8%EB%93%A4%EB%A7%81-%EB%B0%A9%EB%B2%95

realsigridjin commented on src/main/java/com/hyune/raider50g/service/WebClientService.java:
```diff
@@ -0,0 +1,114 @@
+        ? discordMessages
+        : discordMessages.stream()
+            .filter(message -> {
+              User messageAuthor = message.getUserAuthor().orElseGet(DiscordUser::new);
```
Optional 사용이 많네요. 옵셔널 사용은 비싼데요, Dan만의 옵셔널 사용 전략이 있을까요.
http://homoefficio.github.io/2019/10/03/Java-Optional-%EB%B0%94%EB%A5%B4%EA%B2%8C-%EC%93%B0%EA%B8%B0/

realsigridjin replied:
author가 없는데도 새로운 Discord User를 생성하여 반환하고 있네요. 이런 경우에는 분기처리하기 어려울 것입니다.
Null Object Pattern을 적용해보시는 것을 추천해요.

realsigridjin commented on src/main/java/com/hyune/raider50g/util/JsonUtil.java:
```diff
@@ -0,0 +1,34 @@
+
+  static {
+    gson = new GsonBuilder()
+        .registerTypeAdapter(LocalDate.class, new TypeAdapter<LocalDate>() {
```
TypeAdapter가 무엇인가요?

realsigridjin commented on src/main/java/com/hyune/raider50g/util/JsonUtil.java:
```diff
@@ -0,0 +1,34 @@
+          public LocalDate read(final JsonReader jsonReader) throws IOException {
+            return LocalDate.parse(jsonReader.nextString());
+          }
+        }.nullSafe()).create();
```
nullSafe가 뭔가요?

realsigridjin commented on src/main/java/com/hyune/raider50g/web/BookingController.java:
```diff
@@ -0,0 +1,42 @@
+    return bookingService.cancelBooking(bookingCommand);
+  }
+
+  @GetMapping("/invite-macro")
```
REST API의 원칙에 따르면 하나의 명사가 하나의 슬래시에 담겨야 합니다. 또한, 동사는 행위에 엮여있어야 합니다.
제가 추천드리는 것은,
```
("/invite/macro")
```
다음 링크를 참고하세요.
https://sanghaklee.tistory.com/57

Hyune-c replied:
개선해보겠습니다.

realsigridjin commented on src/main/java/com/hyune/raider50g/web/BookingController.java:
```diff
@@ -0,0 +1,42 @@
+  @GetMapping("/invite-macro")
+  public String makeInviteMacro(
+      @RequestParam @DateTimeFormat(pattern = "yyyy-MM-dd") LocalDate raidDate,
+      @RequestParam(required = false) @DateTimeFormat(pattern = "yyyy-MM-dd") String exceptUserName) {
```
필수적으로 존재하지 않아도 되는 필드라면 기본값을 설정해줘서 넘겨주는 것은 어떨까요?
null 반환하는 것보다는 나을 것 같습니다.

realsigridjin commented on src/main/java/com/hyune/raider50g/web/ChannelController.java:
```diff
@@ -0,0 +1,33 @@
+
+  private final ChannelService channelService;
+
+  @GetMapping("/{targetDungeon}")
```
저 같으면, 다음과 같이 REST API 링크를 작성하겠습니다.
```/dungeon/target```

realsigridjin commented on src/test/java/com/hyune/raider50g/repository/BookingRepositoryTest.java:
```diff
@@ -0,0 +1,53 @@
+  private BookingRepository bookingRepository;
+
+  @DisplayName("find")
+  @Nested
```
굳이 ```@Nested```를 사용해야 하나요?
별도의 클래스로 분리하는 것이 낫지 않을까요?

Hyune-c replied:
아래와 같이 테스트를 분류할 수 있기에 `@Nested` 를 사용했습니다.  

![image](https://user-images.githubusercontent.com/55722186/88456864-5d237600-cebc-11ea-930f-35d72a10f0ee.png)

realsigridjin replied:
@Hyune-c 아 저렇게도 쓸 수 있군요. 그렇다면 괜찮네요.

realsigridjin commented on src/test/java/com/hyune/raider50g/service/BookingServiceTest.java:
```diff
@@ -0,0 +1,89 @@
+public class BookingServiceTest {
+
+  @Autowired
+  BookingService bookingService;
```
Mockito 라이브러리를 사용해서 mocking해보는 것도 검토해보세요.

Hyune-c replied:
솔직히.. 한번 해보다가 실패했습니다 ㅠ

realsigridjin commented on src/test/java/com/hyune/raider50g/web/BookingControllerTest.java:
```diff
@@ -0,0 +1,60 @@
+import org.springframework.test.web.servlet.result.MockMvcResultMatchers;
+
+@DisplayName("Booking : Web")
+@WebMvcTest(BookingController.class)
```
컨트롤러 레이어에서 ```@WebMvcTest``` 사용한 것이 좋네요.

realsigridjin commented on src/test/java/com/hyune/raider50g/service/WebClientServiceTest.java:
```diff
@@ -0,0 +1,53 @@
+    @DisplayName("지정된 author 없이")
+    @Test
+    public void withNullAuthor() {
+      // given
```
꼼꼼한 given / when / then 패턴 사용이 좋네요 👍

realsigridjin commented on src/test/java/com/hyune/raider50g/web/ChannelControllerTest.java:
```diff
@@ -0,0 +1,54 @@
+  public void sendBookingList() {
+    // given
+    DungeonType dungeonType = DungeonType.TEST;
+    UriComponentsBuilder uriComponentsBuilder = UriComponentsBuilder.newInstance()
```
WebTestClient의 충분한 기능 사용 좋습니다 👍

realsigridjin commented on src/test/java/com/hyune/raider50g/web/ChannelControllerTest.java:
```diff
@@ -0,0 +1,54 @@
+import org.springframework.test.web.reactive.server.WebTestClient.ResponseSpec;
+import org.springframework.web.util.UriComponentsBuilder;
+
+@RunWith(SpringRunner.class)
```
JUnit 5의 Jupiter API를 사용하면 ```@RunWith(SpringRunner.class)``` 설정을 할 필요가 없습니다.
이 기회에 JUnit이 어떠한 구조로 동작하는 지 살펴보세요.
https://blog.codefx.org/design/architecture/junit-5-architecture-jupiter/

sigridjineth reviewed (approved):
수고 많으셨습니다.
전반적으로 우수한 설계라고 생각됩니다.
이 기회에 ATDD 인수테스트에도 도전해보세요.

realsigridjin commented on src/test/java/com/hyune/raider50g/web/BookingControllerTest.java:
```diff
@@ -0,0 +1,60 @@
+class BookingControllerTest {
+
+  @Autowired
+  private MockMvc mockMvc;
```
MockMvc말고 RestAssured 사용을 저는 많이 추천받았습니다. 아래의 의견은 제 의견은 아닙니다만 적어둡니다.
RestAssured를 사용하면, 예를 들어 Spring Rest Docs에서 mockmvc를 사용한다고 하더라도, 서블릿 컨테이너를 띄우되, RestAssured 문법을 사용할 수 있다고 합니다.
RestAssured라는 테스트 객체는 실제 서버가 띄워진 상태에서 테스트를 할 수 있는데요. mockmvc라는 것을 띄우면 실제 서버가 아닌데요. RestAssured의 경우 실제 말하는 서블릿 컨테이너가 mock 컨테이너를 띄우는 것이므로 가볍고, 세부 설정이 가능합니다. 그래서 mock 서블릿 컨테이너를 사용하는 restassured 사용을 추천 많이 하더라구요.

Hyune-c replied:
`RestAssured` 는 키워드 자체를 처음 들었는데 한번 찾아보겠습니다.

realsigridjin commented:
![image](https://user-images.githubusercontent.com/41055141/88394424-6815d200-cdfa-11ea-8caa-273ee7073e47.png)
### 새로운 방법으로 TDD를 접근
When we’re implementing a feature, we start by writing an acceptance test, which exercises the functionality we want to build. While it’s failing, an acceptance test demonstrates that the system does not yet implement that feature; when it passes, we’re done.

Growing object-oriented software, guided by tests

### 새로운 방법으로 TDD를 접근
우리는 기능을 구현할 때, 만들고자 하는 기능을 수행하는 인수 테스트를 작성하는 것으로 시작한다. 인수 테스트가 실패하는 동안은 시스템이 아직 그 기능을 구현하지 않고 있다는 것을 보여준다; 인수 테스트가 통과되면, 기능 구현은 끝이다.

테스트 주도 개발로 배우는 객체 지향 설계와 실천

jypthemiracle commented:
1. 코드가 잘 읽히는가 & 말이 되는가?
* 저는 삼항 연산자는 지양하는 편입니다. 가독성이 좋지 않아서요. 그 외에는 다 괜찮습니다.
* 디미터 법칙을 잘 지켜주세요.
2. Issue 에 내용이 잘 정리되어있는가?
* 이슈 제목을 명령문으로 작성하시는 것을 추천드립니다. 제목만 봐서는 제 3자가 의도를 파악하기 어렵습니다.
![image](https://user-images.githubusercontent.com/41055141/88396454-84673e00-cdfd-11ea-8fb9-7eddfdea9969.png)
출처: https://helloworld.kurly.com/blog/fix-style-with-command/
3. Repository 가 읽기 쉬운가? (특히 README)
* README.md에는 프로젝트의 소개와 빌드 방법, 기술 스택을 간략하게 적어주세요.
* 구체적으로 진행하신 고민 사항은 깃헙 저장소 위키를 사용해주세요.
우수 사례: https://github.com/codesquad-member-2020/issue-tracker-01
우수 사례2: https://github.com/connect-foundation/2019-07
4. Work Log 가 유용한가?
* 깃헙 위키에 관리하시는 것이 가독성에 좋아보입니다.
우수 사례: https://github.com/connect-foundation/2019-01/wiki/Project-Tech-Log

realsigridjin commented:
1. 코드가 잘 읽히는가 & 말이 되는가?
* 저는 삼항 연산자는 지양하는 편입니다. 가독성이 좋지 않아서요. 그 외에는 다 괜찮습니다.
* 디미터 법칙을 잘 지켜주세요.
2. Issue 에 내용이 잘 정리되어있는가?
* 이슈 제목을 명령문으로 작성하시는 것을 추천드립니다. 제목만 봐서는 제 3자가 의도를 파악하기 어렵습니다.
![image](https://user-images.githubusercontent.com/41055141/88396454-84673e00-cdfd-11ea-8fb9-7eddfdea9969.png)
출처: https://helloworld.kurly.com/blog/fix-style-with-command/
3. Repository 가 읽기 쉬운가? (특히 README)
* README.md에는 프로젝트의 소개와 빌드 방법, 기술 스택을 간략하게 적어주세요.
* 구체적으로 진행하신 고민 사항은 깃헙 저장소 위키를 사용해주세요.
우수 사례: https://github.com/connect-foundation/2019-07
4. Work Log 가 유용한가?
* 깃헙 위키에 관리하시는 것이 가독성에 좋아보입니다.
우수 사례: https://github.com/connect-foundation/2019-01/wiki/Project-Tech-Log

Hyune-c commented:
@jypthemiracle 

이번 리뷰는 이정도에서 끝내고, 다음 버전에서는 좀 더 개선해서 돌아오겠습니다.
리뷰 감사합니다!!
